### PR TITLE
Refactor UniBull

### PR DESCRIPTION
**Pull request by haythem96**
# Task: Refactor UniBull

## Description

- As not every contract that imported UniBull needed both of the TWAP price and flashswap, refactored UniBull into 2 components: UniOracle library and UniFlash contract
- Clean bunch on un-needed code/contracts
- Changing bunch of state vars to immutable 
- Fix deposit

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update

## How Has This Been Tested

Please describe how to test to verify the changes. Provide instructions so we can reproduce.

## FE Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] Added video recordings if it is a UI change

## User Facing Checklist

- [ ] I fully understand the user problem this PR is solving
- [ ] I know who the target user is for this PR and have a deep understanding of that user
- [ ] I have tried this flow thinking from the pov of the target user for this PR
- [ ] I (or working w someone on team) have scheduled a user test for this PR (if it is a large change)

### CODE REVIEW

File: packages/bull-vault/test/e2e/FlashBullTestFork.t.sol [hunk 1]
```diff
@@ -376,8 +327,8 @@ contract FlashBullTestFork is Test {
         uint256 ethInLendingBefore = IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy));
         uint256 usdcBorrowedBefore = IEulerDToken(dToken).balanceOf(address(bullStrategy));
         uint256 crabBalanceBefore = crabV2.balanceOf(address(bullStrategy));
-        uint256 ethToWithdrawFromBull = ethToWithdrawFromCrab.sub(maxEthForSqueeth).add(wethToWithdraw.sub(maxEthForUsdc));
-        userEthBalanceBeforeTx = user1.balance;
+        // uint256 ethToWithdrawFromBull = ethToWithdrawFromCrab.sub(maxEthForSqueeth).add(wethToWithdraw.sub(maxEthForUsdc));
```
Comment (daryakaviani) [inline packages/bull-vault/test/e2e/FlashBullTestFork.t.sol:330]: Should we delete these?

File: packages/bull-vault/src/FlashBull.sol [hunk 2]
```diff
@@ -86,15 +85,14 @@ contract FlashBull is UniBull {
...
         bullStrategy = _bull;
-        oracle = _oracle;
         crab = IBullStrategy(_bull).crab();
-        wPowerPerp = IController(IBullStrategy(bullStrategy).powerTokenController()).wPowerPerp();
-        weth = IController(IBullStrategy(bullStrategy).powerTokenController()).weth();
-        usdc = IController(IBullStrategy(bullStrategy).powerTokenController()).quoteCurrency();
-        ethWSqueethPool = IController(IBullStrategy(bullStrategy).powerTokenController()).wPowerPerpPool();
-        ethUSDCPool = IController(IBullStrategy(bullStrategy).powerTokenController()).ethQuoteCurrencyPool();
+        wPowerPerp = IController(IBullStrategy(_bull).powerTokenController()).wPowerPerp();
+        weth = IController(IBullStrategy(_bull).powerTokenController()).weth();
```
Comment (daryakaviani) [inline packages/bull-vault/src/FlashBull.sol:92]: Maybe we should set a var for the powerTokenController so we don't have to constantly call this function? Which is cheaper?

**Review by daryakaviani — APPROVED** [review]
some small comments

File: packages/bull-vault/test/e2e/FlashBullTestFork.t.sol (hunk 1 above)
Comment (haythem96) [reply packages/bull-vault/test/e2e/FlashBullTestFork.t.sol:330]: let's keep it and try to find it how to do it

File: packages/bull-vault/src/FlashBull.sol [hunk 3]
```diff
@@ -144,10 +135,14 @@ contract FlashBull is UniBull {
             abi.encodePacked(_ethToCrab)
         );
 
-        (uint256 ethInCrab, uint256 squeethInCrab) = IBullStrategy(bullStrategy).getCrabVaultDetails(); 
-        // USDC we pay to flashswap
-
-        (ethToLend, usdcToBorrow) = IBullStrategy(bullStrategy).calcLeverageEthUsdc(crabAmount, share, ethInCrab, squeethInCrab, IERC20(crab).totalSupply());
+        (ethInCrab, squeethInCrab) = IBullStrategy(bullStrategy).getCrabVaultDetails(); 
+        uint256 share;
+        if (IERC20(bullStrategy).totalSupply() == 0) {
```
Comment (aleone) [inline packages/bull-vault/src/FlashBull.sol:140]: I feel like we have some repeated code like this - maybe consider if we can refactor to function or something else to make it cleaner?

File: packages/bull-vault/test/e2e/FlashBullTestFork.t.sol [hunk 4]
```diff
@@ -68,17 +68,12 @@ contract FlashBullTestFork is Test {
...
             eulerMarketsModule
         );
-        uniBullHelper = new UniBullHelper(
-            0x1F98431c8aD98523631AE4a59f267346ea31F984
-        );
         flashBull = new FlashBull(
             address(bullStrategy),
-            0x1F98431c8aD98523631AE4a59f267346ea31F984,
-            0x65D66c76447ccB45dAf1e8044e918fA786A483A1
+            0x1F98431c8aD98523631AE4a59f267346ea31F984
```
Comment (aleone) [inline packages/bull-vault/test/e2e/FlashBullTestFork.t.sol:76]: can we have this be a variable?  its the factory right?

File: packages/bull-vault/test/e2e/FlashBullTestFork.t.sol (hunk 1 above)
Comment (aleone) [reply packages/bull-vault/test/e2e/FlashBullTestFork.t.sol:330]: Made comment in prior pr in how to calculate it - use uniswap quoter simulate a swap and see what you get

**Review by aleone — COMMENTED** [review]
small comments

**Pull request merged**